---
title: "Crypto_popgenModule_md"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("vcfR")
library("adegenet")
library("textplot")
library("wordcloud")
```

## Read in data
Reading in merged vcf files generated using bcftools merge, containing SNP data from multiple samples. Convert this into a genind object, and assign populations according to the prefix of the sample ID*. Finally, convert into a genpop object.

*this is not optimal since it only works in this instance and assumes that all members of UKH26, for example, belong to the same population. This could be switched out for populations as defined by a text file.

```{r}
vcf <- read.vcfR( "/home/amorris/BioInf/M_interjectum_WD/parapipe_run/GVCFs/merged.vcf", verbose = FALSE )
dna <- ape::read.dna("/home/amorris/BioInf/M_interjectum_WD/parapipe_run/REFDATA/mycobacterium_tuberculosis.fasta", format = "fasta")
gff <- read.table("/home/amorris/BioInf/M_interjectum_WD/parapipe_run/REFDATA/mycobacterium_tuberculosis.gff", sep="\t", quote="")
species <- read.table("/home/amorris/BioInf/M_interjectum_WD/species_list.csv", sep=",", header=TRUE)
chrom <- create.chromR(name='Supercontig', vcf=vcf, seq=dna, ann=gff)

gind <- vcfR2genind(vcf)
gind
row.names(gind@tab)
pops_meta <- sapply(strsplit(as.character(indNames(gind)), "_"), "[[", 1)
df <- data.frame(pops = pops_meta, id = indNames(gind))
df <- df[order(df$id),]
gind@pop <- as.factor(df$pops)
gpop <- genind2genpop(gind)
gpop
```

## Plot data
Run centred, non-scaled PCA on data.
```{r, fig.width=25, fig.height=15}
X <- tab(gind, NA.method="mean")
pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
temp <- as.integer(pop(gind))
myCol <- transp(c("blue","red"),.7)[temp]
myPch <- c(15,17)[temp]

plot(pca1$li, col=myCol, cex=3, pch=myPch)

textplot(pca1$li[,1], pca1$li[,2], words=rownames(X), cex=1.2, new=FALSE)

abline(h=0,v=0,col="grey",lty=2)
s.arrow(pca1$c1*.5, add.plot=TRUE)
legend("topright", pch=c(15,17), col=transp(c("blue","red"),.7),
        leg=c("Group A","Group B"), pt.cex=1.2)
```

Generate a dendrogram using distance in Euclidean space
```{r, fig.width=25, fig.height=25}
library(ape)
D <- dist(tab(gind))
stree <- nj(D)

myBoots <- boot.phylo(stree, as.matrix(D), function(x) nj(x), B = 1000,  mc.cores = 6)

p <- ggtree(stree) + geom_treescale()

dd <- as.data.frame(species)

p %<+% dd + geom_tiplab(aes(fill = species) )
  


# par(xpd="True")
# 
# target_tips <- stree$"tip.label"[startsWith(stree$"tip.label", "A19")]
# 
# plot(stree, type="unrooted", edge.w=2, col=c("red")[target_tips])
# edgelabels(tex=round(stree$edge.length,1), bg=rgb(.8,.8,1,.8))
# plot(stree, type="p", edge.w=2)
```
Bootstrapped tree
```{r, fig.width=30, fig.height=15}
library(ape)
library(ggtree)

D <- dist(tab(gind))
stree <- nj(D)
# ggtree(stree, layout = "circular") + geom_tiplab(offset=4, align=FALSE)
par(xpd="True")
plot(stree, type="fan", edge.w=2)

myBoots <- boot.phylo(stree, as.matrix(D), function(x) nj(x), B = 10000,  mc.cores = 6)
edgelabels(tex=myBoots, bg=rgb(.8,.8,1,.8))
plot(stree, type="p", edge.w=2)
edgelabels(tex=myBoots, bg=rgb(.8,.8,1,.8))

plot.phylo(stree, cex = 1, x.lim = 0.3)
nodelabels(myBoots, adj = 0.7, cex=0.7, frame = "none", col = "red")
tiplabels(stree$tip)
```

Carry out Principle coordinate analysis on these euclidean distances, since they are a more accurate representation of nth dimensional distances in planar space.
```{r, fig.width=40, fig.height=25}
pco1 <- dudi.pco(D, scannf=FALSE,nf=2)
s.label(pco1$li*1.1, clab=0, pch="")
textplot(pco1$li[,1], pco1$li[,2], words=rownames(pco1$li),
cex=1.4, new=FALSE, xpd=TRUE)
title("Principal Coordinate Analysis\n-based on SNP distances-")
```

Replace missing data with mean, and generate another PCA. Produce eigenvalue plots and plot PCA using ALL eigenvalues (setting this in an automated manner could be fiddly). Various representations of a PCA plot are given here.
```{r, fig.width=25, fig.height=20}
X <- scaleGen(gind, NA.method="mean")
pca1 <- dudi.pca(X,cent=FALSE,scale=FALSE,scannf=FALSE,nf=3)
barplot(pca1$eig[1:50],main="PCA eigenvalues", col=heat.colors(50))

s.label(pca1$li)
add.scatter.eig(pca1$eig, 3,1,2)
```

Improvement to PCA plot by using s.class to represent both the genotypes and inertia ellipses for populations.
```{r, fig.width=20, fig.height=15}
s.class(pca1$li,pop(gind),xax=1,yax=3,sub="PCA 1-2",csub=2)
title("PCA of SNP dataset\naxes 1-2")
add.scatter.eig(pca1$eig[1:20],nf=3,xax=1,yax=2)
```
and on the third axis
```{r, fig.width=20, fig.height=15}
s.class(pca1$li,pop(gind),xax=1,yax=3,sub="PCA 1-3",csub=2)
title("PCA of SNP dataset\naxes 1-3")
add.scatter.eig(pca1$eig[1:20],nf=3,xax=1,yax=3)
```

Colourisation of the PCA
```{r, fig.width=20, fig.height=15}
col <- funky(25)
s.class(pca1$li, pop(gind),xax=1,yax=3, col=transp(col,1), axesell=FALSE,
cstar=1, cpoint=3, grid=FALSE)
```

Recode principle components to be represented on the RGB scale
```{r, fig.width=10, fig.height=7}
colorplot(pca1$li, pca1$li, transp=TRUE, cex=3, xlab="PC 1", ylab="PC 2")
title("PCA of SNP dataset\naxes 1-2")
abline(v=0,h=0,col="grey", lty=2)
```
and on the third axis
```{r, fig.width=10, fig.height=7}
colorplot(pca1$li[c(1,3)], pca1$li, transp=TRUE, cex=3, xlab="PC 1", ylab="PC 3")
title("PCA of SNP dataset\naxes 1-3")
abline(v=0,h=0,col="grey", lty=2)
```

Kmeans clustering
```{r, fig.width=10, fig.height=7}
# library(tidyverse)  # data manipulation
library(cluster)    # clustering algorithms
library(factoextra) # clustering algorithms & visualization
library(gridExtra)

gind.mat <- as.matrix(gind$tab)
gind.mat[is.na(gind.mat)] <- 0
gind.mat[gind.mat==2] <- 1
filtered.mat <- gind.mat[ , which(apply(gind.mat, 2, var) != 0)]
wss.p <- fviz_nbclust(filtered.mat, kmeans, method = "wss")
grid.arrange(wss.p, nrow = 1)
sc.km <- kmeans(filtered.mat, centers = 2, nstart =5)
p2 <- fviz_cluster(sc.km, data = filtered.mat, labelsize=0, main="k=2")
grid.arrange(p2, nrow = 1)
```